### ISSUE TITLE
ext/XS-APItest/t/locale.t failure in perl 5.39.7 on Windows 8.1 using gcc-13.2.0

### ISSUE THREAD

**Issue by twata1** [Perl/perl5#21880]
**Description**

I have been building Perl 5.39.4 and 5.39.6 on Windows 8.1 64 bit using gcc version 13.2.0 (MinGW-W64 x86_64-msvcrt-posix-seh, built by Brecht Sanders), which can be found on https://winlibs.com/ .
The tests have passed.

This time I am building Perl 5.39.7.
Several types of warnings are output (e.g. `ignoring '#pragma intrinsic ' [-Wunknown-pragmas]`,  `unused parameter 'piPerl' [-Wunused-parameter]`, etc.), but the compilation itself is successful.

When I run `gmake test`, I get a message box at `../ext/XS-APItest/t/locale.t` that says "perl.exe has stopped working".
I have also run `.\perl harness ../ext/XS-APItest/t/locale.t` as follows with similar results.

```
C:\Perl64-gcc13\perl-5.39.7\t>.\perl harness ../ext/XS-APItest/t/locale.t
../ext/XS-APItest/t/locale.t .. Dubious, test returned 255 (wstat 65280, 0xff00)
All 63 subtests passed

Test Summary Report
-------------------
../ext/XS-APItest/t/locale.t (Wstat: 65280 (exited 255) Tests: 63 Failed: 0)
  Non-zero exit status: 255
  Parse errors: No plan found in TAP output
Files=1, Tests=63, 18 wallclock secs ( 0.00 usr +  0.00 sys =  0.00 CPU)
Result: FAIL
Finished test run at Wed Jan 24 01:00:16 2024.

C:\Perl64-gcc13\perl-5.39.7\t>
```

The results of `mode con` are as follows.
```
C:\Perl64-gcc13\perl-5.39.7\t>mode con

Status for device CON:
----------------------
    Lines:          9999
    Columns:        130
    Keyboard rate:  31
    Keyboard delay: 1
    Code page:      1252


C:\Perl64-gcc13\perl-5.39.7\t>
```

**Steps to Reproduce**

Download perl-5.39.7.tar.gz from metacpan (https://metacpan.org/release/CORION/perl-5.39.7).
Extract from perl-5.39.7.tar.gz to C:\Perl64-gcc13.
Launch Command Prompt.
On that Command Prompt, run the following.
_(Note: I use gmake.exe, which is a renamed mingw32-make.exe.)_

```
set PATH=C:\winlibs-x86_64-posix-seh-gcc-13.2.0-llvm-17.0.4-mingw-w64msvcrt-11.0.1-r2\mingw64\bin;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0
set CCHOME=C:\winlibs-x86_64-posix-seh-gcc-13.2.0-llvm-17.0.4-mingw-w64msvcrt-11.0.1-r2\mingw64
cd C:\Perl64-gcc13\perl-5.39.7\win32
gmake
```
After `gmake` is finished , run `gmake test` on the same Command Prompt.

**Expected behavior**

No tests failure.

**Perl configuration**

```
C:\Perl64-gcc13\perl-5.39.7>perl -V
Summary of my perl5 (revision 5 version 39 subversion 7) configuration:

  Platform:
    osname=MSWin32
    osvers=6.3.9600
    archname=MSWin32-x64-multi-thread
    uname=''
    config_args='undef'
    hint=recommended
    useposix=true
    d_sigaction=undef
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=undef
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='gcc'
    ccflags ='-Wall -Wextra -Werror=pointer-arith -Wno-format -Wno-long-long -Werror=vla -std=c99 -DWIN32 -DWIN64 -DPERL_TEXTMODE_SCRIPTS -DMULTIPLICITY -DPERL_IMPLICIT_SYS -DUSE_PERLIO -D__USE_MINGW_ANSI_STDIO -fwrapv -fno-strict-aliasing -mms-bitfields'
    optimize='-Os'
    cppflags='-DWIN32'
    ccversion=''
    gccversion='13.2.0'
    gccosandvers=''
    intsize=4
    longsize=4
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='long long'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='g++'
    ldflags ='-s -L"C:\WinLibs\Perl-5.39.7\lib\CORE" -L"C:\winlibs-x86_64-posix-seh-gcc-13.2.0-llvm-17.0.4-mingw-w64msvcrt-11.0.1-r2\mingw64\lib" -L"C:\winlibs-x86_64-posix-seh-gcc-13.2.0-llvm-17.0.4-mingw-w64msvcrt-11.0.1-r2\mingw64\x86_64-w64-mingw32\lib" -L"C:\winlibs-x86_64-posix-seh-gcc-13.2.0-llvm-17.0.4-mingw-w64msvcrt-11.0.1-r2\mingw64\lib\gcc\x86_64-w64-mingw32\13.2.0"'
    libpth=C:\winlibs-x86_64-posix-seh-gcc-13.2.0-llvm-17.0.4-mingw-w64msvcrt-11.0.1-r2\mingw64\lib C:\winlibs-x86_64-posix-seh-gcc-13.2.0-llvm-17.0.4-mingw-w64msvcrt-11.0.1-r2\mingw64\x86_64-w64-mingw32\lib C:\winlibs-x86_64-posix-seh-gcc-13.2.0-llvm-17.0.4-mingw-w64msvcrt-11.0.1-r2\mingw64\lib\gcc\x86_64-w64-mingw32\13.2.0
    libs= -lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32 -ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr -lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32
    perllibs= -lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32 -ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr -lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32
    libc=-lmsvcrt
    so=dll
    useshrplib=true
    libperl=libperl539.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_win32.xs
    dlext=dll
    d_dlsymun=undef
    ccdlflags=' '
    cccdlflags=' '
    lddlflags='-shared -s -L"C:\WinLibs\Perl-5.39.7\lib\CORE" -L"C:\winlibs-x86_64-posix-seh-gcc-13.2.0-llvm-17.0.4-mingw-w64msvcrt-11.0.1-r2\mingw64\lib" -L"C:\winlibs-x86_64-posix-seh-gcc-13.2.0-llvm-17.0.4-mingw-w64msvcrt-11.0.1-r2\mingw64\x86_64-w64-mingw32\lib" -L"C:\winlibs-x86_64-posix-seh-gcc-13.2.0-llvm-17.0.4-mingw-w64msvcrt-11.0.1-r2\mingw64\lib\gcc\x86_64-w64-mingw32\13.2.0"'



Characteristics of this binary (from libperl):
  Compile-time options:
    HAS_LONG_DOUBLE
    HAS_TIMES
    HAVE_INTERP_INTERN
    MULTIPLICITY
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_HASH_FUNC_SIPHASH13
    PERL_HASH_USE_SBOX32
    PERL_IMPLICIT_SYS
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_USE_SAFE_PUTENV
    USE_64_BIT_INT
    USE_ITHREADS
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
  Built under MSWin32
  Compiled at Jan 24 2024 23:41:02
  @INC:
    C:/Perl64-gcc13/perl-5.39.7/lib

C:\Perl64-gcc13\perl-5.39.7>
```

**Comment by sisyphus** [comment]
Yes, I see the same on Windows 11 using an MSVCRT winlibs gcc-13.2.0 compiler,
However, all tests pass if I use the UCRT variant of the same compiler.
The UCRT compilers have much better locale support than the MSVCRT ones, and I think that's a good reason to switch to the UCRT type.
(With recent devel releases, $Config{libc} will tell you whether you've used UCRT or MSVCRT.)
I've been using the UCRT (13.1.0, then 13.2.0 from winlibs) starting with perl-5.38.0 and have encountered no issues specific to that runtime.

However, I think we still want the perl test suite to pass when the MSVCRT is used.
@khw, what do you think ?

Incidentally, if, in your build folders, you <code>cd</code> to the ext\XS-APItest folder, you can run <code>..\\..\perl -I..\\..\lib t\locale.t</code>.
For my MSVCRT build, running it that way passes all of the tests that get run (up to and including test 63) - but I don't know whether at that point it silently aborts, or whether it has completed successfully.
UPDATE: Duh, if I had read the "Test Summary Report" properly, I would have realized that it has aborted silently.
@twata, what do you see ?

**Comment by twata1** [comment]
@sisyphus 
Actual test reports can be found here.
[gmake-test-2024-01-24.log.txt](https://github.com/Perl/perl5/files/14054434/gmake-test-2024-01-24.log.txt)

The results of `..\..\perl -I..\..\lib t\locale.t` are as follows.
![001-results-locale t](https://github.com/Perl/perl5/assets/20146676/a1f511a1-4d6f-4764-842b-3cd471550993)
![002-results-locale t](https://github.com/Perl/perl5/assets/20146676/5a226406-5f26-4d77-9c45-a9e829497e2f)

**Comment by khwilliamson** [comment]
I was able to reproduce this, and it is a bug I already have fixed locally, and will submit a PR in the next day or two.

Thanks for the report

**Comment by twata1** [comment]
@khwilliamson 
It was worth reporting.
Thank you so much!

**Comment by twata1** [comment]
I have patched the contents of #21891 into the perl-5.39.7 source and can confirm that the issue has been resolved.
The test report is attached.
[gmake-test-2024-02-04.log.txt](https://github.com/Perl/perl5/files/14157310/gmake-test-2024-02-04.log.txt)

Thank you!
